I1099 wti scoreboard problem details

clevengr opened the pull request:
### Description of what the PR does
 Updates the scoreboard on the WTI "Scoreboard" menu tab so that it displays submission details for each problem (instead of just showing "Solved" and "Score").

### Issue which the PR addresses
Partial fix for Issue #1099.   What is still missing is:
- ~~Red/Yellow/Green background colors in each problem cell according to the status of that problem for the corresponding team.~~ Done
- ~~Colored boxes at the top of the individual problem columns (i.e. the problem letters) corresponding to the balloon colors assigned to the problems.~~ Done
- Problem letters (column headers) are currently hard-coded from left to right (A..B..C....).  This will only match the actual problem letter if the problems are loaded into PC2 in alphabetical order.

Adding these things is somewhat problematic because of missing scoring data (see Issue #1100).  This PR does include a partial Fix for some of the problems noted in Issue #1100.

This PR also Fixes #1108.

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Eclipse 2020-12, Windows 11, Java version "11.0.16.1" 2022-08-18 LTS

### Precise steps for _testing_ the PR

- Start a PC2 server using `--load tenprobs` to load the sample contest that has 10 problems.
- Start a PC2 Admin.
- Select the Admin `Problems` tab.  Observe that the last two problems are (sort of) incorrectly named `Sumit a`, whereas logically they should be named `Sumit i` and `Sumit j` for consistency with the rest of the problem set naming.  Edit these two problems and change the name if you like (it isn't strictly necessary, but it will avoid the confusion of having three identically-name problems when you start submitting runs...).
- Select the Admin `Settings` tab.  Scroll down to the "Team Settings" box.  Observe that the configured "Team Scoreboard Display Format" is quite cumbersome (it's confusing when you look at it on the scoreboard).  Recommend changing it to (something like) `{:clientnumber}  {:teamname} Group {:groupid}:{:groupname}`.  This gives a reasonable long yet still readable display. Change the text and then click `Update`.
- Start a PC2 Judge and log in as "judge6" (this is the first judge configured to AutoJudge all the problems).
- Unzip the WTI distribution (in the `projects` folder), then go into the resulting WebTeamInterface-1.2 folder and start the WTI Server, using the command `./bin/pc2wti`.
- Open a browser.  If you have previously used this browser to connect to PC2, _clear the browser's data cache_ (or, use "incognito" mode).
- Connect to the PC2 WTI Server from the browser and login as a team.
- Click the `Scoreboard` menu item.  Observe that every team is displayed and includes a column for each problem (A-J) showing entries `0/--` in all problem fields.  (This means "zero submissions/no-solution-time" and is identical to the format used on the standard PC2 HTML scoreboard).
- Submit an _incorrect_ run for one of the problems.
- Observe that the count of submissions for that team and that problem has increased by one.
- Submit a _correct_ run for the same problem.  (Note that the only _languages_ configured in this contest are Java and Python, so you'll need a solution for the Sumit problem written in one of those two languages (or else you'll need to add a language).  Since all problems are the standard "Sumit" problem, you can find a correct solution in the PC2 `samps/src` folder:  either `ISumit.java` or `isumit.py`.  You'll also need to have a compiler on your machine for the chosen language).
- Observe that 
  - the count of submissions for that team for that problem has increased, 
  - the `--` under the problem has been updated to the time (number of minutes) at which the correct solution was submitted, and
  - the `Solved` column now shows one solution and the `Score` column shows the correct score for the team.
- Try submitting various correct and incorrect runs for other problems; verify that the entries under the corresponding problem, and the Solved and Score entries, update as expected.
- Try logging in as a different team and submitting various correct and incorrect runs; verify that the entries under the problems, and the Solved and Score values, update as expected.

clevengr committed:
i1099: initial version displaying individual problems on scoreboard

clevengr committed:
i1099: first attempt to change background color of prob details

clevengr committed:
i1099: revert scoreboard problem details to no-coloring

clevengr committed:
i1099: center-align some scoreboard elements.

clevengr committed:
i1099: fix cente-alignment in problem details row data.

johnbrvc reviewed (approved):
The only comment I have is that there are a bunch of console.log messages being printed now.  Maybe these should be protected with a debug flag of some sort?  
Anyway, changes look good and are seemingly benign as they relate to the rest of the system.

I installed the WTI on the nac25practice system and it is running now and seems to work.

Commit:
i1099: add scss rules to color problem detail cells - not complete.

clevengr committed:
i1099: support for r/y/g cells with rounded corners.

kkarakas reviewed (commented):
Code looks good.

kkarakas commented:
It is hard to test this on my end since I cannot have a submission accepted.

clevengr committed:
i1099: create shared sccs formatting for yes/no/pending

clevengr committed:
i1099: try ChatGPT-recommended border spacings (not much help...)

clevengr committed:
i1099: update cell spacing

clevengr commented:
> It is hard to test this on my end since I cannot have a submission accepted.

@kkarakas I'm not quite sure what you mean by this.  If you have followed the steps under "Precise steps for testing the PR" -- including you have specified `--load tenprobs` when you started the PC2 Server -- then you have a contest with ten problems in it.  Every one of those ten problems is identical:  each one is the "Sumit" problem from the PC2 `samps` folder.  You can submit an "accepted" solution for any of the problems by using the `isumit` code in `samps/src`.  For example, `isumit.cpp` is a correct C++ solution; `isumit.py` is a correct Python solution; and `Isumit.java` is a correct Java solution (notice the capital "I" on the Java file name).

Please let me know if you still can't get a correct solution submitted.

You might want to pull the latest updates from my branch; they contain code to render submissions with colored cells (red/yellow/green).

clevengr committed:
i1099: correct font, add explicit padding for problem detail cells

clevengr committed:
i1099: increase border spacing in scoreboard table cells

clevengr committed:
i1099: increase rounded-corner radius on problem detail cells

clevengr committed:
i1099: comment-out excessive console logging

kkarakas commented:
@clevengr Thanks for the input. Well the reason I said what you quoted is because I did not set up so that I can run py, cpp etc files locally and check if they output what is expected. I don't know how to do that. Probably I have to download compilers for each language? Now that I think about it I could probably overwrite submission and give an AC although that might defeat the purpose of testing.

clevengr commented:
> @clevengr Thanks for the input. Well the reason I said what you quoted is because I did not set up so that I can run py, cpp etc files locally and check if they output what is expected. I don't know how to do that. Probably I have to download compilers for each language? Now that I think about it I could probably overwrite submission and give an AC although that might defeat the purpose of testing.

@kkarakas  But you have to have Java installed (else how could you compile PC2, right?)  So just submit ISumit.java for "correct" solutions, any other Java program for "incorrect" answers, and any non-Java program for "compilation errors".  That will allow you to generate all possible types of responses.  (Keep in mind that the only thing this PR does is display problem result details -- so you only need to see Correct, Incorrect, and Pending runs.  If you really want to generate specific types of "incorrect" solutions there are Java programs in `samps/src` that will do that as well.)

kkarakas commented:
@clevengr Ah didn't know you can do that, will look into it.

clevengr committed:
i1099: add padding around Letters; set nowrap for problem state.

clevengr committed:
i1099: add scoreboard-table class for uniform spacing

clevengr committed:
i1099: change "pending" color from yellow to "ICPC Blue"

kkarakas commented:
It generally seems to work well. As discussed and noted when there are multiple A's and more than 28 question data spills out. I noticed it takes a bit for the scoreboard to load however that is not a big issue. Did not try the very last commit.

kkarakas commented:
I used the newest version and the colors work well the numbers are visible. I observed another known problem where when the number in the colored box increase the box widens as well.

clevengr committed:
i1099: add support for rendering FTS in darker green.

Commit:
i1099: use custom code instead of BalloonSetting to pass colors to WTI

Commit:
i1099: update problem details header cells to show balloon colors

Commit:
i1099: add support for scoreboard header label URLs to problems PDFs

Commit:
i1099: disable problem writeup header links until contest has started

Commit:
i1099: update 'url' prop in standings output, to "problems/<letter>.pdf"

Commit:
i1099: only show problem detail headers and data if contest is started

clevengr committed:
i1099: add default problem writeups for scoreboard links to WTI-UI

johnbrvc commented:
So, is this ready to be reviewed yet?  That is, no more changes are being made in terms of functionality?

clevengr committed:
i1099: update WebServer to track contest state & prob access.  Includes:

- Defining a new class "ContestStartedServlet". This servlet responds to
POSTs to endpoint "/internal/contestStarted", but ONLY if they come from
localhost.  A POST from localhost to this URL marks the contest as
having been started.
- Installing the new servlet in Jetty.
- Defining a new "ContestState" class encapsulating contest state
   (including started or not) from the WebServer's point of view.
- Adding an instance of ContestState to the WebServer, and passing it to
various Jetty endpoint handlers.
- Wrapping the previously-existing "webApp" handler inside a new class
  "ProtectProblemWritesupsHandler".  This new class rejects accesses to
  "/problems" (where contest problem writeups are stored) if the contest
  has not started; if the contest HAS started it simply forwards any
  webApp request to the previously-existing handler.

clevengr committed:
i1099: update ScoreboardChangeListener to track "contestStarted", incl:

- invokes ContestController.postContestHasStarted() when the SCL sees a
ContestClock event indicating the contest has started.

clevengr committed:
i1099: add HTTP Client to ContestController for contestStarted posts.

clevengr committed:
i1099: add new ContestState class to WTI 'config' package.

clevengr commented:
This PR should have been closed when the initial set of commits were approved and used at NAC25.  However, that didn't happen; additional work was committed after NAC25. A new PR #1123 has been created to replicate (just) the changes which took place _prior_ to NAC 25.  PR #1123 effectively replaces this PR, which is hereby withdrawn and closed without merging.  (See PR #1123 for a more detailed description of the situation.)

A NEW Issue (#1124) and corresponding PR (#tobedetermined) will be submitted to encapsulate the changes which were originally intended to be part of PR #1102 (for Issue #1099) but happened after NAC25.

clevengr commented:
The "unmerged commits" referenced in the GitHub "Closed with unmerged commits" message (below) are all the commits comprising this PR.  Those commits will be transferred to two separate new branches so that they can be merged separately -- the ones used at NAC25 and the ones developed after that.
